### PR TITLE
[Improvement] (pipeline) Cancel related query if backend restarts or dead

### PR DESCRIPTION
**Pull request by zhiqiang-hhhh**
## Proposed changes

This is the 2nd pr of #23704  , now we can cancel query if related backend restarts or dead. Privious pr is #23582

### CODE REVIEW

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.55 seconds
 stream load tsv:          544 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162062480 Bytes

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

File: be/src/agent/heartbeat_server.cpp [hunk 1]
```diff
@@ -81,6 +81,7 @@ void HeartbeatServer::heartbeat(THeartbeatResult& heartbeat_result,
         heartbeat_result.backend_info.__set_be_node_role(config::be_node_role);
         // If be is gracefully stop, then k_doris_exist is set to true
         heartbeat_result.backend_info.__set_is_shutdown(doris::k_doris_exit);
+        heartbeat_result.backend_info.__set_be_process_uuid(_be_epoch);
```
Comment (yiguolei) [inline be/src/agent/heartbeat_server.cpp:84]: why not use start time?

File: gensrc/thrift/HeartbeatService.thrift [hunk 2]
```diff
@@ -50,6 +50,7 @@ struct TBackendInfo {
     6: optional i64 be_start_time
     7: optional string be_node_role
     8: optional bool is_shutdown
+    9: optional i64 be_process_uuid
```
Comment (yiguolei) [inline gensrc/thrift/HeartbeatService.thrift:53]: uuid is  int128, int64 could not be a uuid

File: be/src/agent/heartbeat_server.cpp (hunk 1 above)
Comment (zhiqiang-hhhh) [reply be/src/agent/heartbeat_server.cpp:84]: its it same, be start time is assigned to _be_epoch either.

File: gensrc/thrift/HeartbeatService.thrift (hunk 2 above)
Comment (zhiqiang-hhhh) [reply gensrc/thrift/HeartbeatService.thrift:53]: Rename be_process_uuid to be_process_epoch? Maybe use epoch is enough.

**Comment by zhiqiang-hhhh** [comment]
run FE UT

**Comment by zhiqiang-hhhh** [comment]
run buildadd

**Comment by zhiqiang-hhhh** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java [hunk 3]
```diff
@@ -1252,6 +1253,92 @@ public RowBatch getNext() throws Exception {
...
+
+    // We use a very conservative cancel strategy.
+    // 0. If backends has zero process epoch, do not cancel. Zero process epoch usually arises in cluster upgrading.
+    // 1. If process epoch is same, do not cancel. Means backends does not restart or die.
+    public boolean shouldCancel(List<Backend> currentBackends) {
+        Map<Long, Backend> curBeMap = Maps.newHashMap();
+        for (Backend be : currentBackends) {
+            curBeMap.put(be.getId(), be);
+        }
+        lock();
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java:1265]: use  try (lock()) like RAII in cpp

Comment (zhiqiang-hhhh) [reply fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java:1265]: but here we do not have potential thrown exception? can we use try {lock()} finally {unlock();} either?

File: fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java [hunk 4]
```diff
@@ -1252,6 +1253,92 @@ public RowBatch getNext() throws Exception {
...
+
+        for (Long id : idToBackend.keySet()) {
```
Comment (yiguolei) [inline fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java:1267]: 这里为什么要判断这个idToBackend， 按道理说直接BackendExecStates 就行了吧？

Comment (zhiqiang-hhhh) [reply fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java:1267]: 嗯，这里可以删掉，只用 BackendExecStates 跟 PipelineExecContext 也是可以的

Comment (zhiqiang-hhhh) [reply fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java:1267]: > 这里为什么要判断这个idToBackend， 按道理说直接BackendExecStates 就行了吧？

idToBackend 是用来记录 某个 coordinator 在 prepare阶段时候 所有 be 信息的 snapshot。感觉我们这里索引数据结构太多了，很容易用乱，需要一个 multi_index container 这种数据结构

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Review by yiguolei — APPROVED** [review]
LGTM

**Review by xinyiZzz — APPROVED** [review]
LGTM

**Pull request merged**